sync/statesync/state_syncer.go

```diff
@@ -167,30 +182,36 @@ func (t *stateSync) storageTrieProducer(ctx context.Context) error {
 			return err
 		}
 		// If there are no storage tries, then root will be the empty hash on the first pass.
-		if root != (common.Hash{}) {
-			// acquire semaphore (to keep number of tries in progress limited)
-			select {
-			case t.triesInProgressSem <- struct{}{}:
-			case <-ctx.Done():
-				return ctx.Err()
-			}
+		if root == (common.Hash{}) && !more {
+			close(t.segments)
+			return nil
+		}
 
-			// Arbitrarily use the first account for making requests to the server.
-			// Note: getNextTrie guarantees that if a non-nil storage root is returned, then the
-			// slice of account hashes is non-empty.
-			syncAccount := accounts[0]
-			// create a trieToSync for the storage trie and mark it as in progress.
-			storageTrie, err := NewTrieToSync(t, root, syncAccount, NewStorageTrieTask(t, root, accounts))
-			if err != nil {
-				return err
-			}
-			t.addTrieInProgress(root, storageTrie)
-			storageTrie.startSyncing() // start syncing after tracking the trie as in progress
+		// acquire semaphore (to keep number of tries in progress limited)
+		select {
+		case t.triesInProgressSem <- struct{}{}:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		// Arbitrarily use the first account for making requests to the server.
+		// Note: getNextTrie guarantees that if a non-nil storage root is returned, then the
+		// slice of account hashes is non-empty.
+		syncAccount := accounts[0]
+
+		// create a trieToSync for the storage trie and mark it as in progress.
+		storageTrie, err := NewTrieToSync(t, root, syncAccount, NewStorageTrieTask(t, root, accounts))
+		if err != nil {
+			return err
 		}
-		// if there are no more storage tries, close
-		// the task queue and exit the producer.
+		t.addTrieInProgress(root, storageTrie)
+		if !more {
+			close(t.storageTriesDone)
+		}
+		// start syncing after tracking the trie as in progress
+		storageTrie.startSyncing()
```

Comment:
any reason why we did not call this right after `addTrieInProgress`?

Reply:
this is because `startSyncing` runs concurrent to the current thread, so we should establish the expected invariant that if all pending tries are done and the `storageTriesDone` channel is closed before starting the process.

otherwise there could be an (unlikely) race condition where the trie completes syncing but the `storageTriesDone` channel is still open.

Reply:
hmm but closing the `storageTriesDone` channel also means we close `t.segments` [here](https://github.com/ava-labs/subnet-evm/blob/sync-race/sync/statesync/state_syncer.go#L128), right? and then `startSycing` will panic with same error? is there any other invariant?

Reply:
that code is called when the storage trie finishes syncing, and only if there is no other pending ones (and no new ones to be added).